common/configuration/aws_secrets.go

```diff
@@ -0,0 +1,216 @@
+package configuration
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+
+	"github.com/TBD54566975/ftl/internal/slices"
+
+	"github.com/alecthomas/types/optional"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
+	"github.com/aws/aws-sdk-go-v2/service/secretsmanager/types"
+	"github.com/aws/smithy-go"
+)
+
+const schemeKey = "asm"
+
+// AWSSecrets implements Resolver and Provider for AWS Secrets Manager (ASM).
+//
+// The resolver does a direct/proxy map from a Ref to a URL, module.name <-> asm://module.name and does not access ASM at all.
+type AWSSecrets[R Role] struct {
+	AccessKeyID     string
+	SecretAccessKey string
+	Region          string
+	Endpoint        optional.Option[string]
+}
+
+var _ Resolver[Secrets] = AWSSecrets[Secrets]{}
+var _ Provider[Secrets] = AWSSecrets[Secrets]{}
+var _ MutableProvider[Secrets] = AWSSecrets[Secrets]{}
+
+var (
+	asmOnce   sync.Once
+	asmClient *secretsmanager.Client
```

Comment:
Let's make these into fields of `AWSSecrets`. If the client were truly global (ie. could be used with any set of credentials), then the calculus would be different, but this client is scoped to a specific set of credentials IIUC?

Reply:
You might need to convert all the methods to pointer receivers so the Once doesn't get copied.

Reply:
You'll need to write a constructor for that to be possible.

Reply:
I had it like you suggested and decided against it because of the need of a constructor and pointer methods (trying to match the other impl's), but will change if it makes more sense to you.

Reply:
Makes sense, we can revisit if it becomes a necessity.